File: src/Timer.cpp
#include "Timer.h"

Timer::Timer() : isRunning_(false) {
}

void Timer::start() {
    startTime_ = std::chrono::high_resolution_clock::now();
    isRunning_ = true;
}

double Timer::stop() {
    if (!isRunning_) return 0.0;
    
    endTime_ = std::chrono::high_resolution_clock::now();
    isRunning_ = false;
    
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime_ - startTime_);
    return duration.count() / 1000.0; // Convert to milliseconds
}

double Timer::elapsed() const {
    if (!isRunning_) {
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime_ - startTime_);
        return duration.count() / 1000.0; // Convert to milliseconds
    }
    
    auto currentTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(currentTime - startTime_);
    return duration.count() / 1000.0; // Convert to milliseconds
}

double Timer::elapsedMicroseconds() const {
    if (!isRunning_) {
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime_ - startTime_);
        return static_cast<double>(duration.count());
    }
    
    auto currentTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(currentTime - startTime_);
    return static_cast<double>(duration.count());
}

void Timer::reset() {
    isRunning_ = false;
}

File: PackPlannerCSharp/PackPlanner/Timer.cs
using System.Diagnostics;

namespace PackPlanner;

/// <summary>
/// Utility class for measuring execution time with high precision
/// </summary>
public class Timer
{
    private readonly Stopwatch _stopwatch;
    private bool _isRunning;

    /// <summary>
    /// Initializes a new instance of the Timer class
    /// </summary>
    public Timer()
    {
        _stopwatch = new Stopwatch();
        _isRunning = false;
    }

    /// <summary>
    /// Start timing
    /// </summary>
    public void Start()
    {
        _stopwatch.Restart();
        _isRunning = true;
    }

    /// <summary>
    /// Stop timing and return elapsed time in milliseconds
    /// </summary>
    /// <returns>Elapsed time in milliseconds</returns>
    public double Stop()
    {
        if (!_isRunning) return 0.0;

        _stopwatch.Stop();
        _isRunning = false;

        return _stopwatch.Elapsed.TotalMilliseconds;
    }

    /// <summary>
    /// Get elapsed time in milliseconds without stopping
    /// </summary>
    /// <returns>Elapsed time in milliseconds</returns>
    public double Elapsed
    {
        get
        {
            if (!_isRunning)
            {
                return _stopwatch.Elapsed.TotalMilliseconds;
            }

            return _stopwatch.Elapsed.TotalMilliseconds;
        }
    }

    /// <summary>
    /// Get elapsed time in microseconds
    /// </summary>
    /// <returns>Elapsed time in microseconds</returns>
    public double ElapsedMicroseconds
    {
        get
        {
            if (!_isRunning)
            {
                return _stopwatch.Elapsed.TotalMicroseconds;
            }

            return _stopwatch.Elapsed.TotalMicroseconds;
        }
    }

    /// <summary>
    /// Reset timer
    /// </summary>
    public void Reset()
    {
        _stopwatch.Reset();
        _isRunning = false;
    }

    /// <summary>
    /// Check if timer is currently running
    /// </summary>
    public bool IsRunning => _isRunning;
}
